File: ai25btech11003/matgeo/12.809/codes/main.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

// Function to write matrix to binary file
void write_matrix_to_file(double matrix[2][2], const char* filename) {
    FILE *file = fopen(filename, "w");
    if (file == NULL) {
        printf("Error opening file for writing\n");
        return;
    }

    // Write matrix dimensions first
    int rows = 2, cols = 2;
    fwrite(&rows, sizeof(int), 1, file);
    fwrite(&cols, sizeof(int), 1, file);

    // Write matrix data
    for (int i = 0; i < 2; i++) {
        for (int j = 0; j < 2; j++) {
            fwrite(&matrix[i][j], sizeof(double), 1, file);
        }
    }

    fclose(file);
}

// Function to calculate eigenvalues of 2x2 matrix
void calculate_eigenvalues(double matrix[2][2], double eigenvalues[2]) {
    // For 2x2 matrix A = [[a, b], [c, d]]
    // Characteristic polynomial: λ² - (a+d)λ + (ad-bc) = 0
    // Using quadratic formula: λ = [(a+d) ± √((a+d)² - 4(ad-bc))] / 2

    double a = matrix[0][0];
    double b = matrix[0][1];
    double c = matrix[1][0];
    double d = matrix[1][1];

    double trace = a + d;          // sum of diagonal elements
    double determinant = a*d - b*c; // determinant

    double discriminant = trace*trace - 4*determinant;

    if (discriminant >= 0) {
        eigenvalues[0] = (trace + sqrt(discriminant)) / 2.0;
        eigenvalues[1] = (trace - sqrt(discriminant)) / 2.0;
    } else {
        // Complex eigenvalues - store real part only
        eigenvalues[0] = trace / 2.0;
        eigenvalues[1] = trace / 2.0;
    }
}

int main() {
    // Define the matrix A = [[1, -1], [2, -2]]
    double A[2][2] = {{1, -1}, {2, -2}};

    printf("Matrix A:\n");
    printf("[%.3f %.3f]\n", A[0][0], A[0][1]);
    printf("[%.3f %.3f]\n", A[1][0], A[1][1]);

    // Calculate eigenvalues
    double eigenvalues[2];
    calculate_eigenvalues(A, eigenvalues);

    printf("\nEigenvalues:\n");
    printf("λ₁ = %.3f\n", eigenvalues[0]);
    printf("λ₂ = %.3f\n", eigenvalues[1]);

    // Write matrix to binary file
    write_matrix_to_file(A, "main.dat");

    // Write eigenvalues to a separate binary file for verification
    FILE *eigen_file = fopen("eigenvalues.dat", "w");
    if (eigen_file != NULL) {
        fwrite(eigenvalues, sizeof(double), 2, eigen_file);
        fclose(eigen_file);
    }

    printf("\nMatrix data written to main.dat\n");
    printf("Eigenvalues written to eigenvalues.dat\n");

    return 0;
}

File: ai25btech11003/matgeo/12.809/codes/main.py
import numpy as np
import struct
import os

def read_matrix_from_file(filename):
    """Read matrix from binary file created by C program"""
    try:
        with open(filename, 'rb') as file:
            # Read dimensions
            rows = struct.unpack('i', file.read(4))[0]
            cols = struct.unpack('i', file.read(4))[0]

            # Read matrix data
            matrix = np.zeros((rows, cols))
            for i in range(rows):
                for j in range(cols):
                    value = struct.unpack('d', file.read(8))[0]
                    matrix[i][j] = value

            return matrix
    except FileNotFoundError:
        print(f"Error: {filename} not found. Please run the C program first.")
        return None

def calculate_eigenvalues_python(matrix):
    """Calculate eigenvalues using numpy"""
    eigenvalues = np.linalg.eigvals(matrix)
    return eigenvalues

def verify_eigenvalues_manual(matrix):
    """Verify eigenvalues using manual calculation for 2x2 matrix"""
    a, b = matrix[0, 0], matrix[0, 1]
    c, d = matrix[1, 0], matrix[1, 1]

    # Characteristic polynomial: λ² - (a+d)λ + (ad-bc) = 0
    trace = a + d
    determinant = a*d - b*c

    # Using quadratic formula
    discriminant = trace**2 - 4*determinant

    if discriminant >= 0:
        lambda1 = (trace + np.sqrt(discriminant)) / 2
        lambda2 = (trace - np.sqrt(discriminant)) / 2
    else:
        # Complex eigenvalues
        real_part = trace / 2
        imag_part = np.sqrt(-discriminant) / 2
        lambda1 = complex(real_part, imag_part)
        lambda2 = complex(real_part, -imag_part)

    return [lambda1, lambda2]

def main():
    print("=== EIGENVALUE CALCULATION - PYTHON SOLUTION ===\n")

    # Check if C program has been run
    if not os.path.exists('main.dat'):
        print("Error: main.dat not found!")
        print("Please run the C program first using:")
        print("gcc -shared -fPIC -o main.so main.c -lm")
        print("gcc -o main main.c -lm")
        print("./main")
        return

    # Read matrix from binary file
    matrix = read_matrix_from_file('main.dat')

    if matrix is not None:
        print("Matrix A (read from main.dat):")
        print(matrix)
        print()

        # Method 1: Using numpy
        eigenvalues_numpy = calculate_eigenvalues_python(matrix)
        print("Eigenvalues using numpy.linalg.eigvals():")
        for i, eigenval in enumerate(eigenvalues_numpy):
            print(f"λ{i+1} = {eigenval:.6f}")
        print()

        # Method 2: Manual calculation
        eigenvalues_manual = verify_eigenvalues_manual(matrix)
        print("Eigenvalues using manual calculation:")
        for i, eigenval in enumerate(eigenvalues_manual):
            if isinstance(eigenval, complex):
                print(f"λ{i+1} = {eigenval}")
            else:
                print(f"λ{i+1} = {eigenval:.6f}")
        print()

        # Verify the characteristic polynomial
        print("Verification - Characteristic Polynomial:")
        a, b = matrix[0, 0], matrix[0, 1]
        c, d = matrix[1, 0], matrix[1, 1]
        trace = a + d
        det = a*d - b*c
        print(f"det(A - λI) = λ² - ({trace})λ + ({det}) = λ² + λ + 0 = λ(λ + 1)")
        print("Setting λ(λ + 1) = 0:")
        print("λ = 0 or λ = -1")
        print()

        print("SOLUTION VERIFICATION:")
        print("The eigenvalues of the given matrix A = [[1, -1], [2, -2]] are:")
        print("λ₁ = 0")
        print("λ₂ = -1")
        print("\nAnswer: Option (a) -1 and 0 is CORRECT")

if __name__ == "__main__":
    main()